src/dual.jl

```diff
@@ -1,68 +1,61 @@
-immutable Dual{T<:Real} <: Number
-    re::T
-    du::T
+typealias ReComp Union{Real,Complex}
+
+immutable Dual{T<:ReComp} <: Number
+    value::T
+    epsilon::T
 end
-Dual(x::Real, y::Real) = Dual(promote(x,y)...)
-Dual(x::Real) = Dual(x, zero(x))
+Dual{S<:ReComp,T<:ReComp}(x::S, y::T) = Dual(promote(x,y)...)
+Dual(x::ReComp) = Dual(x, zero(x))
 
-const du = Dual(false, true)
+const ɛ = Dual(false, true)
 
 typealias Dual128 Dual{Float64}
-typealias Dual64 Dual{Float32}
-typealias DualPair Dual
-
-real(z::Dual) = z.re
-epsilon(z::Dual) = z.du
-
-eps(z::Dual) = eps(real(z))
-eps{T}(::Type{Dual{T}}) = eps(T)
-one(z::Dual) = dual(one(real(z)))
-one{T}(::Type{Dual{T}}) = dual(one(T))
-@deprecate inf{T}(::Type{Dual{T}}) convert(Dual{T}, Inf)
-@deprecate nan{T}(::Type{Dual{T}}) convert(Dual{T}, NaN)
-isnan(z::Dual) = isnan(real(z))
-
-convert{T<:Real}(::Type{Dual{T}}, x::Real) =
-  Dual{T}(convert(T, x), convert(T, 0))
-convert{T<:Real}(::Type{Dual{T}}, z::Dual{T}) = z
-convert{T<:Real}(::Type{Dual{T}}, z::Dual) =
-  Dual{T}(convert(T, real(z)), convert(T, epsilon(z)))
-
-convert{T<:Real}(::Type{T}, z::Dual) =
-  (epsilon(z)==0 ? convert(T, real(z)) : throw(InexactError()))
-
-promote_rule{T<:Real, S<:Real}(::Type{Dual{T}}, ::Type{Dual{S}}) =
-    Dual{promote_type(T, S)}
-# these promotion rules shouldn't be used for scalar operations -- they're slow
-promote_rule{T<:Real}(::Type{Dual{T}}, ::Type{T}) = Dual{T}
-promote_rule{T<:Real, S<:Real}(::Type{Dual{T}}, ::Type{S}) =
-  Dual{promote_type(T, S)}
-
-dual(x, y) = Dual(x, y)
-dual(x) = Dual(x)
-
-@vectorize_1arg Real dual
-@vectorize_2arg Real dual
-@vectorize_1arg Dual epsilon
-
-dual128(x::Float64, y::Float64) = Dual{Float64}(x, y)
-dual128(x::Real, y::Real) = dual128(float64(x), float64(y))
-dual128(z) = dual128(real(z), epsilon(z))
-dual64(x::Float32, y::Float32) = Dual{Float32}(x, y)
-dual64(x::Real, y::Real) = dual64(float32(x), float32(y))
-dual64(z) = dual64(real(z), epsilon(z))
-
+typealias Dual64  Dual{Float32}
+typealias Dual32  Dual{Float16}
+typealias DualComplex256 Dual{Complex128}
+typealias DualComplex128 Dual{Complex64}
+typealias DualComplex64  Dual{Complex32}
+
+convert{T<:ReComp}(::Type{Dual{T}}, z::Dual{T}) = z
+convert{T<:ReComp}(::Type{Dual{T}}, z::Dual) = Dual{T}(convert(T, value(z)), convert(T, epsilon(z)))
+convert{T<:ReComp}(::Type{Dual{T}}, x::Number) = Dual{T}(convert(T, x), convert(T, 0))
+convert{T<:ReComp}(::Type{T}, z::Dual) = (epsilon(z)==0 ? convert(T, value(z)) : throw(InexactError()))
+
+promote_rule{T<:ReComp, S<:ReComp}(::Type{Dual{T}}, ::Type{Dual{S}}) = Dual{promote_type(T, S)}
+promote_rule{T<:ReComp, S<:ReComp}(::Type{Dual{T}}, ::Type{S}) = Dual{promote_type(T, S)}
+promote_rule{T<:ReComp}(::Type{Dual{T}}, ::Type{T}) = Dual{T}
+
+widen{T}(::Type{Dual{T}}) = Dual{widen(T)}
+
+value(z::Dual) = z.value
+epsilon(z::Dual) = z.epsilon
+value(z::Number) = z
```

Comment:
I'd be in favor of removing the no-op until someone requests it for a good reason, but it's just a minor point.